src/test/java/lambda/part3/exercise/ReachIterableTest.java

```diff
@@ -46,32 +46,41 @@
             );
         }
 
-        default <R> ReachIterable<R> flatMap(final Function<T, List<R>> function) {
-            final List<R> results = new ArrayList<>();
-            while (this.forNext(element -> results.addAll(function.apply(element))));
-            return ReachIterable.from(results);
+        default <R> ReachIterable<R> flatMap(final Function<T, List<R>> mapper) {
+            final AtomicReference<ReachIterable<R>> inner = new AtomicReference<>(
```

Comment:
Your code is not thread safe any way, so you could use arrays of size 1 instead of atomics.